Custom Block Icons

mtias opened the pull request:
_In Progress_

This update seeks to add custom icons to the block library in order to reduce the expectation of dashicon usage on blocks, which has resulted in a loss of clarity as soon as multiple blocks share the same icon (quote and pullquote, for example). Dashicons were also designed for UI affordance while blocks have distinct qualities that we should reflect.

It also hopes to encourage people to supply a custom SVG for their own blocks.

Closes #8719.

![image](https://user-images.githubusercontent.com/548849/44029945-77afa140-9eff-11e8-86da-0df7d6408043.png)

The new icons have a 24px base grid (pulling from Material design). All core blocks have been replaced except for a few where we had custom icons already which need updating to the new sizing (shortcode and custom HTML). This also reworks all the icons used for embeds and favors service specific icons when available for clarity.

## Missing Icons

- [x] Shortcode.
- [x] HTML.
- [x] Code.
- [x] Headings.
- [x] Inline Image.
- [x] Cover Image.

## Remaining Tasks

- [x] Address slash inserter.
- [x] Implement stacked icon deck.
- [x] Work on hover styles.
- [ ] Define spread of background color.

mtias committed:
Custom icons for Blocks

Uses custom svgs at a 24px base, 36px upscale for the inserter.
WordPress should signal that block icons ought to be custom for the
sake of clarity for users.

mtias committed:
Update some more icons.

mtias committed:
Add icons for several core blocks.

mtias committed:
Use service icons for embeds.

mtias committed:
Adjust embed icons.

mtias committed:
Expand width of inserter, use lighter design.

mtias committed:
Use ligher line arrow for panel collapse / expand.

mtias committed:
Adjust inserter design giving more paddings.

mtias committed:
Update more block icons and use color API for embeds.

mtias committed:
Adjust svg sizes for non-dashicon icons.

mtias committed:
Adjust block icon stack.

mtias committed:
Use different icons for embed fallback.

mtias committed:
Add more embed logos (WordPress, Tumblr).

chrisvanpatten commented:
I love that this uses raw SVGs instead of some kind of special icon name thing. Encouraging raw SVGs is great.

Hopefully this also opens the door to improving those icons without worrying about matching an icon suite necessarily — that quote icon, for instance, feels like it could use a little love. Because it's raw SVG I imagine it'll be easier to make small case-by-case tweaks rather than trying to get Google to approve a design change upstream.

Super cool!

youknowriad commented:
<img width="404" alt="screen shot 2018-08-13 at 13 03 37" src="https://user-images.githubusercontent.com/272444/44030659-5aea0ec0-9ef9-11e8-95b9-2cefa39fd6df.png">

I think it's not easy to distinguish whether the title applies to the icon above it or under it. I think we can fix this by reducing the padding of the buttons and increasing the margin bottom of the items. (Something like the screenshot above)

Also, I don't like the fact that the scrollbars show up surrounded by a padding. I think it's better to avoid the global padding and move it to each child div instead, that way the scrollbars show up at the edge.

mtias committed:
Add remaining icons for code, shortcode, inline image.

paulwilde commented on packages/block-library/src/image/index.js:
```diff
@@ -89,12 +89,30 @@ const schema = {
+		</g>
+	</svg>
+);
+
```
This SVG seems to have a bunch of unnecessary IDs and <g> tags.

mtias committed:
Simplify image SVG and add new Heading icon.

ccprog commented:
I just came across this PR and noted that some of the SVGs use the following pattern:

    <svg width="24" height="24"><path fill="none" d="M0 0h24v24H0V0z"/>...

Note that there is no `viewBox` attribute, but instead an empty "backround" path element covering the whole viewport. If an icon defined like this is embeded as a `<img>`, its size can be styled. But if it is used inlined in HTML, changing the size with CSS does not work. You would be well-advised to use

    <svg viewBox="0 0 24 24">...

for all icons to take advantage of SVGs intrinsic responsiveness, both to enable theming and for the sake of maintainablility. (All you'll ever have to do is adjust sizes in CSS.)

As an aside none of the other attributes sprinkled through there are really needed:

- `width` and `height` should be and are provided by CSS,
- `version` is never evaluated by browsers,
- `enable-background` is deprecated,
- `xmlSpace` is a valid attribute - in JSX -, but is superfluous for these icons, as they have no text content
- `xmlns="http://www.w3.org/2000/svg"` is a matter of convention. It is not strictly needed for `<svg>` elements inlined in HTML, but if you ever copy-and-paste them to a file that is evaluated as XML (like a standalone SVG file), it is compulsory.

chrisvanpatten commented:
It might be helpful to run the icons through `svgo`, which can help normalise some of the issues that @ccprog mentioned. I usually use this nice GUI option: https://jakearchibald.github.io/svgomg/

ccprog commented:
@chrisvanpatten SVGO is a heavily opinonated tool. In this case, it would _remove_ the viewBox as a default. Despite what the GUI seems to claim, it cannot _add_ a missing viewBox. Option "Prefer viewBox to width/height" only means it will dellete width/height if a viewBox is present. If it is not, a conscious designer decision is needed.

jasmussen committed:
Polish inline image & block chip

This changes the icon for inline images to be bigger, and consistent with the image block.

It also polishes the block chip to ensure it has the same size in the inspector and in the toolbar, and have the right size.

jasmussen commented:
Pushed a little polish to the block chip, and the inline image icon:

![screen shot 2018-08-14 at 10 07 49](https://user-images.githubusercontent.com/1204802/44080015-0d2876c4-9fab-11e8-9e3b-0851b07bc8ab.png)

youknowriad commented:
What impact this PR should have on the block switcher's icon and the block style variations panel? (which mimic the inserter design)

mtias committed:
Cleanup SVG declarations.

jasmussen committed:
Fix appender icons

jasmussen committed:
Normalize SVGs and fix slash command

jasmussen commented:
Pushed a few fixes:

![screen shot 2018-08-14 at 14 37 05](https://user-images.githubusercontent.com/1204802/44093121-df7a844c-9fd2-11e8-9858-94207ae95b27.png)

![screen shot 2018-08-14 at 14 56 40](https://user-images.githubusercontent.com/1204802/44093123-e13abac2-9fd2-11e8-8a87-ba9d5cfa56cb.png)

jasmussen committed:
Normalize focus style in library.

jasmussen commented:
Normalized focus style:

![screen shot 2018-08-14 at 15 06 44](https://user-images.githubusercontent.com/1204802/44093447-dff5c548-9fd3-11e8-8bf0-05829c1e2938.png)

mtias committed:
Fix test issues.

mtias committed:
Clean "Classic" icon and linter issues.

jasmussen committed:
Polish focus, hover styles.

jasmussen commented:
In https://github.com/WordPress/gutenberg/pull/8916/commits/5874f5ba395df9eff02ebc55c7038506b14866d4 I pushed a little polish to focus and hover styles of blocks:

![iterations](https://user-images.githubusercontent.com/1204802/44150861-1435425c-a0a1-11e8-81c7-41098ada0224.gif)

mtias commented:
Oh, better structure for the scrollbar would also be needed.

jasmussen commented:
I can fix it tomorrow so the scrollbar is right on the edge of the window as opposed to floating in the middle, if that's what you mean? I'll be sure to fix the padding also.

Commit:
Tweak how scrolling area works, and add grid units.

jasmussen commented:
I pushed a change to the scrollable area:

<img width="444" alt="screen shot 2018-08-16 at 08 29 55" src="https://user-images.githubusercontent.com/1204802/44192275-26208380-a12f-11e8-8c13-84b71e4be30d.png">

<img width="485" alt="screen shot 2018-08-16 at 08 30 24" src="https://user-images.githubusercontent.com/1204802/44192277-27ea4700-a12f-11e8-9329-470db683352c.png">

Note I also changed the spacing from using 13px which didn't really fit with anything, to using 16px. I stored that as a new grid variable ($grid-size-large, with $grid-size being 8px), which we can use elsewhere to gain consistency.

Commit:
It's padding, not margin. Sigh.

jasmussen commented:
<img width="487" alt="screen shot 2018-08-16 at 11 13 59" src="https://user-images.githubusercontent.com/1204802/44199870-8cb09c00-a145-11e8-9874-0f7deaa4852e.png">

Okay, I think I _actually_ fixed the scrollbar.

Commit:
Fix switcher

jasmussen commented:
Fixed the switcher as well:

<img width="421" alt="screen shot 2018-08-16 at 11 36 08" src="https://user-images.githubusercontent.com/1204802/44201092-adc6bc00-a148-11e8-8ebd-c08e68779f7d.png">

Commit:
Fix style variations.

jasmussen commented:
Fixed style variations and focus styles.

<img width="484" alt="screen shot 2018-08-16 at 11 49 23" src="https://user-images.githubusercontent.com/1204802/44201849-95f03780-a14a-11e8-9f95-2fa9bd2ba8eb.png">

youknowriad committed:
Fix unit tests

paulwilde commented:
> Fixed the switcher as well:

The label text (Quote) looks too far to the bottom of the hover background in this image.

chrisvanpatten commented:
I definitely like the direction this is going but all that padding on the outside edge just feels weird vs having it extend full-width. It creates a weirdly raggedy edge…

![_untitled____add_new_page_ _mindful_ _wordpress](https://user-images.githubusercontent.com/1231306/44208819-90d1bf00-a12f-11e8-9a71-ee5ab69f865b.png)

vs

![_untitled____add_new_page_ _mindful_ _wordpress](https://user-images.githubusercontent.com/1231306/44208796-7dbeef00-a12f-11e8-9177-5c8fb8f6012b.png)

But I do like the new icons (feels super weird for WordPress to use Material Icons instead of a WordPress-specific icon set, but they still work) and the lack of a background on the default icons isn't as bothersome in practice as I thought it might be. Overall looks good!

youknowriad committed:
Refactoring block icons rendering

youknowriad commented:
I pushed a refactoring to BlockIcons rendering which hopefully 🤷‍♂️  does:

 - Fix adding embed blocks (always use `BlockIcon` to render icons)
 - Centralize the CSS to apply colors in `BlockIcon` using the `showColors` prop
 - Consistent rendering for `BlockIcon` with or without colors (always have a `.editor-block-icon` wrapper)
 - Remove contrast checking for block icons (fix e2e tests)

youknowriad committed:
Fix style lint

youknowriad committed:
Remove debug code :) Sorry

youknowriad committed:
Fix icon styles

youknowriad committed:
Fix block icons e2e test

youknowriad commented:
So tests pass, I think we're ready to ship. Could anyone test everything for last minute breakage?

@chrisvanpatten I'm not designer but I had the same thoughts, I feel without the padding, it's better but that's shouldn't block us here. It's still a huge improvement over what we have, we can merge and iterate.

youknowriad reviewed (approved):
LGTM 👍

chrisvanpatten commented:
Only remaining thing I noticed, but shouldn't be a blocker, is that dashicons now render blurry in non-retina displays (they may also be blurry in Retina, but I expect not because pixels!) because they're upscaled. It'd be nice if there was a fallback that just padded the dashicons instead of upscaling. But again, that shouldn't be a blocker.

jorgefilipecosta commented on test/e2e/specs/block-icons.test.js:
```diff
@@ -130,7 +130,7 @@ describe( 'Correctly Renders Block Icons on Inserter and Inspector', () => {
 			await searchForBlock( blockTitle );
 			validateSvgIcon( await getFirstInserterIcon() );
-			expect( await getBackgroundColor( INSERTER_ICON_SELECTOR ) ).toEqual( 'rgb(1, 0, 0)' );
+			expect( await getBackgroundColor( INSERTER_ICON_SELECTOR ) ).toEqual( 'rgba(0, 0, 0, 0)' );
```
The background color is in fact rgb(1, 0, 0) and with this change we don't verify if the background was in fact applied. I think we may need to update the selector const INSERTER_ICON_SELECTOR = `${ INSERTER_BUTTON_SELECTOR } .editor-block-icon`; to be  `${ INSERTER_BUTTON_SELECTOR } .editor-block-types-list__item-icon`.

youknowriad replied:
Oh gotcha I'll take a look

jorgefilipecosta replied:
.editor-block-icon is a child of .editor-block-types-list__item-icon and it seems colors are applied in .editor-block-types-list__item-icon.

youknowriad replied:
It should be fixed.

youknowriad committed:
Make sure the e2e icons test validates the colors properly

karmatosed reviewed (approved):
I think we should ship what we have right now. The indent visually does help for padding on scrolling. Let's get in as we have now and we can always iterate.

youknowriad commented:
Awesome work everyone, I love those PRs where so many people push code and review :P

aduth commented on packages/editor/src/components/block-icon/index.js:
```diff
@@ -25,18 +25,25 @@ function renderIcon( icon ) {
-				{ renderedIcon }
-			</div>
-		);
+	const style = showColors ? {
```
Wasted effort to assign variable when following condition shortcuts early return without using (should be assigned after condition).

aduth reviewed (commented):
Curious the impact this might have on bundle size. Probably not critical impact, but I'm sure it's a bump.

afercia commented:
With this change, all the new hardcoded SVG icons miss the attributes required for accessibility:
`role="img" aria-hidden="true" focusable="false"` 😞Will open a new issue.

jasmussen commented:
You can assign that issue to me.
